invokeai/backend/model_manager/config.py

```diff
@@ -0,0 +1,328 @@
+    """Model config for standalone VAE models."""
+
+    format: Literal[ModelFormat.Checkpoint] = ModelFormat.Checkpoint
+
+
+class VaeDiffusersConfig(ModelConfigBase):
+    """Model config for standalone VAE models (diffusers version)."""
+
+    format: Literal[ModelFormat.Diffusers] = ModelFormat.Diffusers
+
+
+class ControlNetDiffusersConfig(DiffusersConfig):
+    """Model config for ControlNet models (diffusers version)."""
+
+    format: Literal[ModelFormat.Diffusers] = ModelFormat.Diffusers
+
+
+class ControlNetCheckpointConfig(CheckpointConfig):
+    """Model config for ControlNet models (diffusers version)."""
+
+    format: Literal[ModelFormat.Checkpoint] = ModelFormat.Checkpoint
+
+
+class TextualInversionConfig(ModelConfigBase):
+    """Model config for textual inversion embeddings."""
+
+    format: Literal[ModelFormat.EmbeddingFile, ModelFormat.EmbeddingFolder]
+
+
+class MainConfig(ModelConfigBase):
+    """Model config for main models."""
+
+    vae: Optional[str] = Field(None)
+    variant: ModelVariantType = ModelVariantType.Normal
+    ztsnr_training: bool = False
+
+
+class MainCheckpointConfig(CheckpointConfig, MainConfig):
+    """Model config for main checkpoint models."""
+
+    # Note that we do not need prediction_type or upcast_attention here
+    # because they are provided in the checkpoint's own config file.
+
+
+class MainDiffusersConfig(DiffusersConfig, MainConfig):
+    """Model config for main diffusers models."""
+
+    prediction_type: SchedulerPredictionType = SchedulerPredictionType.Epsilon
+    upcast_attention: bool = False
+
+
+class ONNXSD1Config(MainConfig):
+    """Model config for ONNX format models based on sd-1."""
+
+    format: Literal[ModelFormat.Onnx, ModelFormat.Olive]
+    prediction_type: SchedulerPredictionType = SchedulerPredictionType.Epsilon
+    upcast_attention: bool = False
+
+
+class ONNXSD2Config(MainConfig):
+    """Model config for ONNX format models based on sd-2."""
+
+    format: Literal[ModelFormat.Onnx, ModelFormat.Olive]
+    # No yaml config file for ONNX, so these are part of config
+    prediction_type: SchedulerPredictionType = SchedulerPredictionType.VPrediction
+    upcast_attention: bool = True
+
+
+class IPAdapterConfig(ModelConfigBase):
+    """Model config for IP Adaptor format models."""
+
+    format: Literal[ModelFormat.InvokeAI]
+
+
+class CLIPVisionDiffusersConfig(ModelConfigBase):
+    """Model config for ClipVision."""
+
+    format: Literal[ModelFormat.Diffusers]
+
+
+class T2IConfig(ModelConfigBase):
+    """Model config for T2I."""
+
+    format: Literal[ModelFormat.Diffusers]
+
+
+AnyModelConfig = Union[
+    MainCheckpointConfig,
+    MainDiffusersConfig,
+    LoRAConfig,
+    TextualInversionConfig,
+    ONNXSD1Config,
+    ONNXSD2Config,
+    VaeCheckpointConfig,
+    VaeDiffusersConfig,
+    ControlNetDiffusersConfig,
+    ControlNetCheckpointConfig,
+    IPAdapterConfig,
+    CLIPVisionDiffusersConfig,
+    T2IConfig,
+]
```

Comment:
This could be improved by using a discriminated union.

Some benefits are listed here: https://docs.pydantic.dev/latest/api/standard_library_types/#discriminated-unions-aka-tagged-unions

In our case, I think we would want to use the `typing.Annotated` syntax. So it would look something like this:

```python
class LoRAConfig(ModelConfigBase):
    """Model config for LoRA/Lycoris models."""
    type: Literal[ModelType.Lora] = ModelType.Lora
    format: Literal[ModelFormat.Lycoris, ModelFormat.Diffusers]


class VaeCheckpointConfig(ModelConfigBase):
    """Model config for standalone VAE models."""
    type: Literal[ModelType.Vae] = ModelType.Vae
    format: Literal[ModelFormat.Checkpoint] = ModelFormat.Checkpoint

AnyModelConfig = Annotated[
    Union[
        MainCheckpointConfig,
        MainDiffusersConfig,
        LoRAConfig,
        TextualInversionConfig,
        ONNXSD1Config,
        ONNXSD2Config,
        VaeCheckpointConfig,
        VaeDiffusersConfig,
        ControlNetDiffusersConfig,
        ControlNetCheckpointConfig,
        IPAdapterConfig,
        CLIPVisionDiffusersConfig,
        T2IConfig,
    ],
    Field(discriminator='type')
]

```

I'm pretty sure this would also enable us to get rid of the `_class_map` in `ModelConfigFactory`. We'd have to discrimate on the `BaseModelType` as well to distinguish between `ONNXSD1Config` and `ONNXSD2Config`. See the [Nested Discriminated Union example](https://docs.pydantic.dev/latest/api/standard_library_types/#nested-discriminated-unions), which is very similar to our use case.

Reply:
This works and is very elegant. The only problem is that I now get the following FastAPI error on web server load:
```
Traceback (most recent call last):
  File "/home/lstein/invokeai-main/.venv/bin/invokeai-web", line 5, in <module>
    from invokeai.app.api_app import invoke_api
  File "/home/lstein/Projects/InvokeAI/invokeai/app/api_app.py", line 41, in <module>
    from .api.routers import (
  File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 704, in decorator
    self.add_api_route(
  File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 643, in add_api_route
    route = route_class(
  File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 489, in __init__
    self.dependant = get_dependant(path=self.path_format, call=self.endpoint)
  File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/fastapi/dependencies/utils.py", line 261, in get_dependant
    type_annotation, depends, param_field = analyze_param(
  File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/fastapi/dependencies/utils.py", line 372, in analyze_param
    assert field_info is None, (
AssertionError: Cannot specify FastAPI annotations in `Annotated` and default value together for 'info'
```
I don't know where to look to find the source of this issue. Could you help?

Reply:
I think this might actually be a bug in FastAPI with this logic: https://github.com/tiangolo/fastapi/blob/480620372a662aa9025c47410fbc90a255b2fc94/fastapi/dependencies/utils.py#L372

I think it is trying to protect against someone doing this:
```
arg: Annotated[int, Body(description="foo")] = Body(description="bar")
```
But it is getting a false alarm on our use of `Annotated`.

Haven't tested, but I think we could work around it like this:
```python
class AnyModelConfig(BaseModel):
    config: Annotated[
        Union[
            _MainModelConfig,
            _ONNXConfig,
            _VaeConfig,
            _ControlNetConfig,
            LoRAConfig,
            TextualInversionConfig,
            IPAdapterConfig,
            CLIPVisionDiffusersConfig,
            T2IConfig,
        ],
        Field(discriminator="type"),
    ]
```

It's a little unfortunate, but I think the extra layer of indirection is worth it for the stronger typing (if this works).

Reply:
This is a fairly longstanding issue with FastAPI. I've run into it before, can't remember my solution tho.

I think using `Body` is actually OK for our purposes. It subclasses `FieldInfo` so it's not too odd, and this is the solution most are using.

Maybe a [custom root model](https://docs.pydantic.dev/latest/api/root_model/) would also work? This does add indirection and a possibility for gotchas (as does what @RyanJDick suggested).

Related issues:
- https://github.com/tiangolo/fastapi/discussions/9761
- https://github.com/tiangolo/fastapi/discussions/9287

Reply:
Unfortunately my proposed solution of using `Body(discriminator="type")` in the discriminated Union works fine from the command line and allows the web server to load, but fails when I try to load `localhost:9090/docs`. Here's the error:

```
  File "/home/lstein/Projects/InvokeAI/invokeai/app/api_app.py", line 124, in custom_openapi
    openapi_schema = get_openapi(
  File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/fastapi/openapi/utils.py", line 530, in get_openapi
    return jsonable_encoder(OpenAPI(**output), by_alias=True, exclude_none=True)  # type: ignore
  File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/pydantic/main.py", line 164, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 8 validation errors for OpenAPI
components.schemas.invokeai__app__api__routers__model_records__ModelsList.Schema.properties.models.Schema.items.Schema.discriminator.mapping.controlnet
  Input should be a valid string [type=string_type, input_value={'discriminator': {'mappi...lNetCheckpointConfig'}]}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.4/v/string_type
components.schemas.invokeai__app__api__routers__model_records__ModelsList.Schema.properties.models.Schema.items.Schema.discriminator.mapping.main
  Input should be a valid string [type=string_type, input_value={'discriminator': {'mappi...MainCheckpointConfig'}]}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.4/v/string_type
components.schemas.invokeai__app__api__routers__model_records__ModelsList.Schema.properties.models.Schema.items.Schema.discriminator.mapping.onnx
  Input should be a valid string [type=string_type, input_value={'discriminator': {'mappi...chemas/ONNXSD2Config'}]}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.4/v/string_type
components.schemas.invokeai__app__api__routers__model_records__ModelsList.Schema.properties.models.Schema.items.Schema.discriminator.mapping.vae
  Input should be a valid string [type=string_type, input_value={'discriminator': {'mappi.../VaeCheckpointConfig'}]}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.4/v/string_type
components.schemas.invokeai__app__api__routers__model_records__ModelsList.Schema.properties.models.Schema.items.bool
  Input should be a valid boolean [type=bool_type, input_value={'description': 'Any of t...ts/schemas/T2IConfig'}]}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.4/v/bool_type
components.schemas.invokeai__app__api__routers__model_records__ModelsList.Schema.properties.models.Schema.items.`list[union[...,bool]]`
  Input should be a valid list [type=list_type, input_value={'description': 'Any of t...ts/schemas/T2IConfig'}]}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.4/v/list_type
components.schemas.invokeai__app__api__routers__model_records__ModelsList.Schema.properties.models.bool
  Input should be a valid boolean [type=bool_type, input_value={'items': {'description':...odels', 'type': 'array'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.4/v/bool_type
components.schemas.invokeai__app__api__routers__model_records__ModelsList.Reference.$ref
  Field required [type=missing, input_value={'description': 'Return l...List', 'type': 'object'}, input_type=dict]
```
I can get rid of the discriminator in `config.py` and add the discriminator back when the route is defined. When I do this, both the command line and the FastAPI interfaces do work:

In `model_records.py`:
```
async def add_model_record(
    config: Annotated[AnyModelConfig,  Body(description="Model config", discriminator="type")]
) -> AnyModelConfig:
```

In `config.py`:
```
_ONNXConfig = Annotated[Union[ONNXSD1Config, ONNXSD2Config], Field(discriminator="base")]
_ControlNetConfig = Annotated[
    Union[ControlNetDiffusersConfig, ControlNetCheckpointConfig], Field(discriminator="format")
]
_VaeConfig = Annotated[Union[VaeDiffusersConfig, VaeCheckpointConfig], Field(discriminator="format")]
_MainModelConfig = Annotated[Union[MainDiffusersConfig, MainCheckpointConfig], Field(discriminator="format")]

AnyModelConfig = Union[
    _MainModelConfig,
    _ONNXConfig,
    _VaeConfig,
    _ControlNetConfig,
    LoRAConfig,
    TextualInversionConfig,
    IPAdapterConfig,
    CLIPVisionDiffusersConfig,
    T2IConfig,
]
```
I've tested what happens from the command line, and it appears that pydantic validation is trying each of the members of the union until it finds a good field match. As far as I can tell everything works fine, but there is a slight performance penalty. Using the discriminated union I can validate a raw dict 100,000 times in 0.2s. Without the discriminated union, this rises to 0.88s . This is a very small performance penalty: roughly 6 millionth of a second per fetch.

I've committed this workaround; let me know what you think. If there are other solutions that do not involve an extra level of indirection (e.g. `model_info = mm.get_record(key)` `config = model_info.config`) I'd prefer to do this right.

Reply:
I'm happy with you're current implementation, @lstein . Maybe just link the two issues that @psychedelicious found for future reference. Hopefully they get solved upstream at some point 🤞

Reply:
Great. I added a note to `config.py` explaining the issue and pointing to the FastAPI bug reports. I am going ahead and resolving the previous conversations. When you have a chance please give this an approval and I will ask other code owners to approve as well so that we can merge.

Reply:
Pydantic 2.5, just released, introduces a [new way to handle discriminators via callback](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-callable-discriminator). These "callable" discriminators offer an alternative way to describe the union here (won't fix the FastAPI bug tho).

I don't think it's necessarily clearer or better than the current strategy, which nicely maps the classes to concepts.

Reply:
I'm just leaving it be for now. I'm still coming up to speed on Pydantic 2.4.2.